src/librustc_mir/transform/validate.rs

```diff
@@ -81,6 +85,104 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
             self.fail(location, format!("encountered jump to invalid basic block {:?}", bb))
         }
     }
+
+    /// Check if src can be assigned into dest.
+    /// This is not precise, it will accept some incorrect assignments.
+    fn mir_assign_valid_types(&self, src: Ty<'tcx>, dest: Ty<'tcx>) -> bool {
+        if src == dest {
+            // Equal types, all is good.
+            return true;
+        }
+        // Normalize projections and things like that.
+        // FIXME: We need to reveal_all, as some optimizations change types in ways
+        // that require unfolding opaque types.
+        let param_env = self.param_env.with_reveal_all();
+        let src = self.tcx.normalize_erasing_regions(param_env, src);
+        let dest = self.tcx.normalize_erasing_regions(param_env, dest);
+        // It's worth checking equality again.
+        if src == dest {
+            return true;
+        }
+
+        // Type-changing assignments can happen when subtyping is used. While
+        // all normal lifetimes are erased, higher-ranked types with their
+        // late-bound lifetimes are still around and can lead to type
+        // differences. So we compare ignoring lifetimes.
+        struct LifetimeIgnoreRelation<'tcx> {
```

Comment:
I probably want to share this with `interpret/eval_context.rs`... any suggestions for where to put that relator then?

Reply:
I guess I could make this a public function here and import that in the interpreter.

Reply:
you could create a `relate` module in `rustc_mir`. I already dislike that we have `transform/const_prop` accessing `interpret`, but the other way around would be even worse imo.

Reply:
Having validation methods in `validate` makes sense though, IMO.

I could also move this to `util`, of course.

Reply:
most of `validate` should probably not live in `transform` anymore

Reply:
anyway, that's for a different PR. let's merge this as it is right now